Add support for receiving and dispatching NotificationResponse messages

alex-xnor opened the pull request:
This is an initial stab at trying to fix #28.  PostgresNIO already has sufficient support for sending `LISTEN` and `NOTIFY` commands, through e.g. `simpleQuery`, but lacks support for handling the `NotificationResponse` messages that result when a notification comes through.  This PR adds support for handling `NotificationResponse` messages by adding an additional channel handler before `PostgresRequestHandler`, that filters out `NotificationResponse` messages (since `PostgresRequest`s are usually ill-prepared to handle them) and diverts them to handlers registered by channel in a `PostgresNotificationHandlerMap` that can be accessed on `PostgresConnection`.  I'm not tied to the current API exposed here; I could imagine a couple different alternatives, but this seemed like the simplest thing that could possibly work and looks reasonably ergonomic to use.

alex-xnor commented:
That does look like a nicer API.  If it turns into a `PostgresRequest` though, wouldn't that mean you couldn't run other commands while a listen is active?  I understand that some other protocols (namely Redis's) have that restriction, but Postgres doesn't.  I can imagine use cases where you'd want to fetch data from the database in response to a notification, for example, which sounds like it would be difficult to do if implemented this way if I'm understanding correctly.  (You could use another connection but that seems like kind of a heavy-weight solution.)

tanner0101 commented:
Hmm... that's a good point. Maybe LISTEN / NOTIFY could bypass `PostgresRequest` entirely then. Each `LISTEN` / `conn.listen(...)` could append a channel handler to the pipeline listening for notification responses on that channel. When the listen is cancelled, the handler is removed. As long as those handlers are before the `PostgresRequestHandler` and consume notification responses then we wouldn't need to touch anything there.

tanner0101 commented:
For example, this is how `conn.requestTLS()` works: https://github.com/vapor/postgres-nio/blob/master/Sources/PostgresNIO/Connection/PostgresConnection%2BRequestTLS.swift

alex-xnor commented:
Thanks for the feedback.  I've rebased onto the fixed master branch for the CI fixes from #62, switched to a `listen`-with-closure interface for registering notification response handlers, and switched to adding channel handlers dynamically.  In order to handle multiple notification handlers for a given channel, a handler does not consume the notification response, so the request handler was nevertheless modified to ignore notification responses.  `LISTEN` and `UNLISTEN` queries are not generated currently (still leaving that to the application) as it's not clear to me that generating SQL (incl. escaping the channel names as `LISTEN` and `UNLISTEN` would require) is in scope for this project.

tanner0101 commented on Sources/PostgresNIO/Connection/PostgresConnection+Notifications.swift:
```diff
@@ -0,0 +1,56 @@
+        let listenContext = PostgresListenContext()
+        let channelHandler = PostgresNotificationHandler(channel: channel, notificationHandler: notificationHandler, listenContext: listenContext)
+        let pipeline = self.channel.pipeline
+        _ = pipeline.addHandler(channelHandler, name: nil, position: .before(requestHandler))
```
Since `PostgresRequestHandler` is ignoring `.notificationResponse` now, I think you should be able to simply append the `PostgresNotificationHandler`. That would make thing simpler not needing to pass around `requestHandler`.

tanner0101 replied:
Nit: Please use `self.` whenever you access a property (pretend you're in an escaping closure all the time).

tanner0101 commented on Sources/PostgresNIO/Connection/PostgresConnection+Notifications.swift:
```diff
@@ -0,0 +1,56 @@
+    typealias InboundOut = PostgresMessage
+
+    let channel: String
+    let notificationHandler: (PostgresMessage.NotificationResponse, PostgresListenContext) -> Void
```
NIO methods tend to pass `context` first. It might be nice to follow that pattern.

tanner0101 commented on Sources/PostgresNIO/Connection/PostgresConnection+Notifications.swift:
```diff
@@ -0,0 +1,56 @@
+                    notificationHandler(notification, listenContext)
+                }
+            } catch let error {
+                errorCaught(context: context, error: error)
```
`errorCaught` will put the error on the pipeline and eventually [close the connection](https://github.com/vapor/postgres-nio/blob/master/Sources/PostgresNIO/Connection/PostgresConnection%2BConnect.swift#L48). I'm not sure that's what we want here. It might be better to just log the error to the connection's logger.

tanner0101 commented on Sources/PostgresNIO/Connection/PostgresConnection+Notifications.swift:
```diff
@@ -0,0 +1,56 @@
+import NIO
+
+public final class PostgresListenContext {
```
Anything `public` should have docblocks.

alex-xnor commented:
Thanks for the additional feedback. I believe I've addressed all your comments. Please take another look.

duncangroenewald commented:
Are there any code examples of using this ?

akirchhoff-modular commented:
> Are there any code examples of using this ?

It has been a very long time since I have touched this code, but the added tests do show how this can be used. These tests remain present in the current version of the code:

https://github.com/vapor/postgres-nio/blob/9f84290f4f7ba3b3edb749d196243fc2df6b82e6/Tests/IntegrationTests/PostgresNIOTests.swift#L111-L230

Hopefully this helps. I am not likely to be of much further use due to the length of time it has been since I worked on this, and I no longer work on anything that uses this.
